File: backend/src/main/java/com/ourride/model/entity/Ride.java
package com.ourride.model.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "rides", indexes = {
    @Index(name = "idx_rides_user_id", columnList = "userId"),
    @Index(name = "idx_rides_driver_id", columnList = "driverId"),
    @Index(name = "idx_rides_status", columnList = "status"),
    @Index(name = "idx_rides_created_at", columnList = "createdAt")
})
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Ride {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @Column(name = "user_id", nullable = false)
    private Long userId;
    
    @Column(name = "driver_id")
    private Long driverId;
    
    @Column(name = "pickup_latitude", nullable = false, precision = 10, scale = 8)
    private BigDecimal pickupLatitude;
    
    @Column(name = "pickup_longitude", nullable = false, precision = 11, scale = 8)
    private BigDecimal pickupLongitude;
    
    @Column(name = "pickup_address", nullable = false, columnDefinition = "TEXT")
    private String pickupAddress;
    
    @Column(name = "destination_latitude", nullable = false, precision = 10, scale = 8)
    private BigDecimal destinationLatitude;
    
    @Column(name = "destination_longitude", nullable = false, precision = 11, scale = 8)
    private BigDecimal destinationLongitude;
    
    @Column(name = "destination_address", nullable = false, columnDefinition = "TEXT")
    private String destinationAddress;
    
    @Column(name = "ride_type", nullable = false, length = 20)
    @Enumerated(EnumType.STRING)
    private RideType rideType;
    
    @Column(nullable = false, length = 20)
    @Enumerated(EnumType.STRING)
    @Builder.Default
    private RideStatus status = RideStatus.PENDING;
    
    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal price;
    
    @Column(name = "distance_km", precision = 10, scale = 2)
    private BigDecimal distanceKm;
    
    @Column(name = "estimated_duration_minutes")
    private Integer estimatedDurationMinutes;
    
    @Column(name = "scheduled_datetime")
    private LocalDateTime scheduledDateTime;
    
    @Column(name = "started_at")
    private LocalDateTime startedAt;
    
    @Column(name = "completed_at")
    private LocalDateTime completedAt;
    
    @Column(name = "cancelled_at")
    private LocalDateTime cancelledAt;
    
    @Column(name = "cancellation_reason", columnDefinition = "TEXT")
    private String cancellationReason;
    
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;
    
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;
    
    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }
    
    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
    
    public enum RideType {
        CAR, CAR_XL, CAR_PLUS, MOTORBIKE, SCOOTER
    }
    
    public enum RideStatus {
        PENDING,
        DRIVER_ASSIGNED,
        DRIVER_EN_ROUTE,
        ARRIVED,
        IN_PROGRESS,
        COMPLETED,
        CANCELLED
    }
}

File: mobile/client/lib/features/rides/domain/entities/ride.dart
import 'package:equatable/equatable.dart';
import 'driver.dart';

class Ride extends Equatable {
  final int id;
  final int userId;
  final int? driverId;
  final double pickupLatitude;
  final double pickupLongitude;
  final String pickupAddress;
  final double destinationLatitude;
  final double destinationLongitude;
  final String destinationAddress;
  final String rideType; // CAR, CAR_XL, CAR_PLUS
  final String status; // PENDING, DRIVER_ASSIGNED, etc.
  final double price;
  final double? distanceKm;
  final int? estimatedDurationMinutes;
  final DateTime? scheduledDateTime;
  final DateTime? startedAt;
  final DateTime? completedAt;
  final DateTime? cancelledAt;
  final String? cancellationReason;
  final DateTime createdAt;
  final DateTime updatedAt;
  final Driver? driver;

  const Ride({
    required this.id,
    required this.userId,
    this.driverId,
    required this.pickupLatitude,
    required this.pickupLongitude,
    required this.pickupAddress,
    required this.destinationLatitude,
    required this.destinationLongitude,
    required this.destinationAddress,
    required this.rideType,
    required this.status,
    required this.price,
    this.distanceKm,
    this.estimatedDurationMinutes,
    this.scheduledDateTime,
    this.startedAt,
    this.completedAt,
    this.cancelledAt,
    this.cancellationReason,
    required this.createdAt,
    required this.updatedAt,
    this.driver,
  });

  @override
  List<Object?> get props => [
        id,
        userId,
        driverId,
        pickupLatitude,
        pickupLongitude,
        pickupAddress,
        destinationLatitude,
        destinationLongitude,
        destinationAddress,
        rideType,
        status,
        price,
        distanceKm,
        estimatedDurationMinutes,
        scheduledDateTime,
        startedAt,
        completedAt,
        cancelledAt,
        cancellationReason,
        createdAt,
        updatedAt,
        driver,
      ];
}

